macaroni/motorControl/motorControl.ino

```diff
@@ -1,124 +1,215 @@
 #include <LiquidCrystal.h>
-
 #include <Servo.h>
 
 void motor(int val);
 void steer(int val);
 void update();
-void limitDesiredValues(int& requestedSpeed, int& requestedHeading);
+int limitDesiredSpeed(int desiredSpeed);
+int limitDesiredHeading(int desiredHeading);
 void updateHeading();
 void updateSpeed();
-bool getMessage();
-
-//LiquidCrystal lcd(12, 11, 10, 9, 8, 7);
-
-static int LED = 13;
+void tick();
+boolean getMessage();
 
-//Variables for motors & state variables
-Servo EscMotor;
-static int escPin = 10;
-Servo SteeringMotor;
-static int SteeringMotorPin=11;
+static float currentSpeed = 0;
+static float currentHeading = 0;
+static float desiredHeading = 0;
 
-static int currentSpeed = 0;
-static int desiredSpeed=0;
-static int currentHeading = 0;
-static int desiredHeading = 0;
+static volatile int currentTicks = 0; //volatile data for manipulation in interrupt routines
+static          int lastTicks = 0;
+static int ticks_per_rotation = 200;
+static int meters_per_rotation = 0.1107 * 3.14; 
 
 //control limits
 static const int maxSpeed = 30; // maximum velocity 
 static const int minSpeed = -15;
 static const int minSteer = -25;
 static const int maxSteer = 25;
 
-void setup() 
+static const int maxHeading = 1;
+static const int minHeading = 1;
+
+static float   pid_p = 0.1;
+static float   pid_i = 0.1;
+static float   pid_d = 0.1;
+
+#define HISTORY_SIZE 100
+static float         desiredSpeed = 0;
+static float         errorSum = 0;
+static float         errorHistory[HISTORY_SIZE] = {0};
+static unsigned long lastSpeedUpdateMicros = 0;
+static int           historyIndex = 0;
+
+const int estopPin = A1;
+boolean estop = false;
+static boolean timeout = false;
+
+Servo esc;
+const int escMux = 10;
+Servo steering;
+const int steerMux = 11;
+
+const int encoderA = 2;
+const int encoderB = 4;
+
+const int muxStatePin = A0;
+boolean muxState = false;
+
+volatile int tickData = 0;
+
+static int lastMessageTime;
+void setup()
 {
-  Serial.begin(115200);
-  EscMotor.attach(escPin);
-  SteeringMotor.attach(SteeringMotorPin);
-  motor(0);
-  steer(0);
-//  lcd.begin(16, 2);
-  // Print a message to the LCD.
-//  lcd.print("Steer    Speed");
-  pinMode(LED, OUTPUT);
-  digitalWrite(LED, LOW);
-} 
-
-unsigned long time = 0;
-
-void loop() {
-  
-  update();
-//  lcd.setCursor(0, 1);
-//  lcd.print("               ");
-//  lcd.setCursor(0, 1);
-//  lcd.print(desiredHeading);
-//  lcd.setCursor(9,1);
-//  lcd.print(desiredSpeed);
-}
+    pinMode(estop, INPUT);
+
+    pinMode(escMux, OUTPUT);
+    pinMode(steerMux, OUTPUT);
+
+    pinMode(encoderA, INPUT);
+    attachInterrupt(digitalPinToInterrupt(encoderA), tick, CHANGE);
+    pinMode(encoderB, INPUT);
+
+    pinMode(muxState, INPUT);
+
+    digitalWrite(escMux, 0);
```

Comment:
Consider giving these a better name, and all writes to these ports should go through the Servo API (ie esc.write(90) to make sure it's giving the stop command). Also, the Servo objects are never attached to these pins...